libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -62,9 +64,236 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
+                    break;
+                case AP_GPS::GPS_Status::NO_FIX:
+                    _fix_state.status = uavcan::equipment::gnss::Fix::STATUS_NO_FIX;
+                    break;
+                case AP_GPS::GPS_Status::GPS_OK_FIX_2D:
+                    _fix_state.status = uavcan::equipment::gnss::Fix::STATUS_2D_FIX;
+                    break;
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D:
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_DGPS:
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_RTK_FLOAT:
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_RTK_FIXED:
+                    _fix_state.status = uavcan::equipment::gnss::Fix::STATUS_3D_FIX;
+                    break;
+            }
+
+            switch(cgps.status()) {
+                case AP_GPS::GPS_Status::NO_GPS:
+                case AP_GPS::GPS_Status::NO_FIX:
+                case AP_GPS::GPS_Status::GPS_OK_FIX_2D:
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D:
+                    _fix2_state.mode = uavcan::equipment::gnss::Fix2::MODE_SINGLE;
+                    _fix2_state.sub_mode = uavcan::equipment::gnss::Fix2::SUB_MODE_DGPS_OTHER;
+                    break;
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_DGPS:
+                    _fix2_state.mode = uavcan::equipment::gnss::Fix2::MODE_DGPS;
+                    _fix2_state.sub_mode = uavcan::equipment::gnss::Fix2::SUB_MODE_DGPS_OTHER;
+                    break;
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_RTK_FLOAT:
+                    _fix2_state.mode = uavcan::equipment::gnss::Fix2::MODE_RTK;
+                    _fix2_state.sub_mode = uavcan::equipment::gnss::Fix2::SUB_MODE_RTK_FLOAT;
+                    break;
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_RTK_FIXED:
+                    _fix2_state.mode = uavcan::equipment::gnss::Fix2::MODE_RTK;
+                    _fix2_state.sub_mode = uavcan::equipment::gnss::Fix2::SUB_MODE_RTK_FIXED;
+                    break;
+            }
+
+            _fix_state.gnss_time_standard = uavcan::equipment::gnss::Fix::GNSS_TIME_STANDARD_UTC;
+            _fix2_state.gnss_time_standard = _fix_state.gnss_time_standard;
+            uavcan::Timestamp ts;
+            ts.usec = 1000ULL * ((cgps.time_week() * AP_MSEC_PER_WEEK + cgps.time_week_ms()) + UNIX_OFFSET_MSEC);
+            _fix_state.gnss_timestamp = ts;
+            _fix2_state.gnss_timestamp = ts;
+
+            _fix_state.num_leap_seconds = 0;
+            _fix2_state.num_leap_seconds = 0;
+            ts.usec = 0;
+            _fix_state.timestamp = ts;
+            _fix2_state.timestamp = ts;
+
+            _fix_out_sem->give();
+        } else {
+            if (sem_ret) {
+                _fix_out_sem->give();
+            }
+        }
+    }
+
+    if (attitude_out_array[_uavcan_i] != nullptr) {
+        bool sem_ret = _att_out_sem->take(1);
+
+        if (sem_ret) {
```

Comment:
this second sem_ret variable is unnecessary (its never used, the take can be called within the if) and just adds risk of confusing it with the first usage of it.